sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java

```diff
@@ -0,0 +1,211 @@
+    RubyVisitor rubyVisitor = (RubyVisitor) ((RubyClass) klazz).allocate();
+    rubyVisitor.metaDataProvider = args[0].toJava(TreeMetaDataProvider.class);
+    return rubyVisitor;
+  }
+
+  @JRubyMethod
+  public IRubyObject process(ThreadContext context, IRubyObject node) {
+    // some nodes present in the AST but are empty (e.g. function parameters clause)
+    // or node is nil (absent)
+    if (noLocation(node) || node.isNil()) {
+      return new RubyNil(getRuntime());
+    }
+
+    IRubyObject processedNode = callSuperMethod(context, "process", node);
+
+    if (processedNode instanceof NodeAdapter) {
+      // A more specific tree was already created by a specialized method (on_${type}, ex: on_int)
+      return processedNode;
+
+    } else {
+      return createNativeTreeAdapter(processedNode);
+    }
+  }
+
+  @JRubyMethod(name = "on_class")
+  public IRubyObject onClass(ThreadContext context, IRubyObject node) {
+    IRubyObject updatedNode = callSuperMethod(context, "on_class", node);
+
+    IdentifierTree identifier = (IdentifierTree) ((NodeAdapter) getChild(updatedNode, 0)).getTree();
+    return convertToNodeAdapter(updatedNode, metaData ->
+      new ClassDeclarationTreeImpl(metaData, identifier, createNativeTree(updatedNode)));
+  }
+
+  @JRubyMethod(name = "on_const")
+  public IRubyObject onConst(ThreadContext context, IRubyObject node) {
+    // FIXME add scope node child to current node
+    Object name = getChild(node, 1);
+    return convertToNodeAdapter(node, metaData -> new IdentifierTreeImpl(metaData, String.valueOf(name)));
+  }
+
+  @JRubyMethod(name = "on_int")
+  public IRubyObject onInt(ThreadContext context, IRubyObject node) {
+    Object value = getChild(node, 0);
+    return convertToNodeAdapter(node, metaData -> new LiteralTreeImpl(metaData, String.valueOf(value)));
+  }
+
+  private NodeAdapter convertToNodeAdapter(IRubyObject node, Function<TreeMetaData, Tree> getTree) {
+    TreeMetaData metaData = getMetaData(node);
+    return NodeAdapter.create(getRuntime(), node, getTree.apply(metaData));
+  }
+
+  private NodeAdapter createNativeTreeAdapter(IRubyObject processedNode) {
+    return NodeAdapter.create(getRuntime(), processedNode, createNativeTree(processedNode));
+  }
+
+  private Tree createNativeTree(IRubyObject node) {
+    TreeMetaData metaData = getMetaData(node);
+    return new NativeTreeImpl(metaData, new RubyNativeKind(nodeType(node).asJavaString()), getChildren(node, metaData));
+  }
+
+  private boolean noLocation(IRubyObject node) {
+    if (!node.isNil()) {
+      IRubyObject location = (IRubyObject) JavaEmbedUtils.invokeMethod(getRuntime(), node, "location", null, IRubyObject.class);
+      SourceMapAdapter sourceMapAdapter = new SourceMapAdapter(getRuntime(), location);
+      RangeAdapter rangeAdapter = sourceMapAdapter.getRange();
+
+      return rangeAdapter.isNull();
+    }
+
+    return false;
+  }
+
+  private TreeMetaData getMetaData(IRubyObject node) {
```

Comment:
Can this be moved to `NodeAdapter` ?

Reply:
I'll leave this for now as is, and we'll improve that in another PR